Lifetime cleanups

nnethercote opened the pull request:
The last commit is very opinionated, let's see how we go.

r? @oli-obk

nnethercote committed:
Remove unnecessary lifetime from `UnsafeInferVarsVisitor`.

nnethercote committed:
Rename some lifetimes.

Giving them more typical names.

nnethercote committed:
Remove unnecessary lifetime from `FeaturePreviouslyDeclared`.

nnethercote committed:
Remove unnecessary lifetime from `BuiltinTypeAliasBounds`.

nnethercote committed:
Remove unnecessary lifetimes from `Arena`.

nnethercote committed:
Rename and reorder lots of lifetimes.

- Replace non-standard names like 's, 'p, 'rg, 'ck, 'parent, 'this, and
  'me with vanilla 'a. These are cases where the original name isn't
  really any more informative than 'a.
- Replace names like 'cx, 'mir, and 'body with vanilla 'a when the lifetime
  applies to multiple fields and so the original lifetime name isn't
  really accurate.
- Put 'tcx last in lifetime lists, and 'a before 'b.

nnethercote committed:
Remove unnecessary lifetime from `CheckLoopVisitor`.

nnethercote committed:
Remove unnecessary lifetimes from `rustc_expand`.

nnethercote committed:
Remove unnecessary lifetime from `RunCompiler`.

This does change the external interface, but not in a way that will
cause any breakage because external users don't mention the lifetimes.

oli-obk commented on compiler/rustc_passes/src/loops.rs:
```diff
@@ -64,8 +64,8 @@ impl fmt::Display for BreakContextKind {
-struct CheckLoopVisitor<'a, 'tcx> {
-    sess: &'a Session,
+struct CheckLoopVisitor<'tcx> {
+    sess: &'tcx Session,
```
We can get the session from the tcx, the field can be removed

nnethercote replied:
I've added a commit doing that.

nnethercote committed:
Remove `sess` from `CheckLoopVisitor`.

lcnr reviewed (commented):
r=me after oli's review.

I feel like this change is right at the edge of whether it's worth it given the potential merge conflicts, reviewer time and what not. So while I think it's an improvement and we should merge this PR, I would not like to review many more PRs like this.

nnethercote commented:
> more PRs like this

The good news is that this PR does most of the these sorts of possible changes :)

lcnr commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 359b6589ac2efdbfedf42f29098394c1f62d4dd3 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 359b6589ac2efdbfedf42f29098394c1f62d4dd3 -->
<!-- homu: {"type":"Approved","sha":"359b6589ac2efdbfedf42f29098394c1f62d4dd3","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
FWIW, how I got to this PR:
- I noticed some convoluted memory management in `rustc_mir_transform`. I cleaned this up in #129720. One part of this change reduced the number of lifetimes for `FilterInformation` from four to two.
- This got me looking for other unnecessary lifetimes. In #130022 I removed unnecessary lifetimes from multiple widely-used structs in dataflow/borrowck code.
- I also noticed that `rustc_resolver` was using `'a` for all references into the `ResolverArenas` type. This is a situation that is perfect for a more descriptive name, so in #130208 I introduced the `'ra` lifetime name.
- This PR is the tail end of this work, where I found a few more unnecessary lifetimes and fixed up some lifetime ordering and naming inconsistencies. I didn't address every single possible inconsistency, just the more obvious ones.
